controllers/volumereplicationgroup_webhook.go

```diff
@@ -0,0 +1,141 @@
+// SPDX-FileCopyrightText: The RamenDR authors
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	ramen "github.com/ramendr/ramen/api/v1alpha1"
+	recipe "github.com/ramendr/recipe/api/v1alpha1"
+	authentication "k8s.io/api/authentication/v1"
+	authorization "k8s.io/api/authorization/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+//nolint:lll
+//+kubebuilder:webhook:path=/validate-ramendr-openshift-io-v1alpha1-volumereplicationgroup,mutating=false,failurePolicy=fail,sideEffects=None,groups=ramendr.openshift.io,resources=volumereplicationgroups,verbs=create;update,versions=v1alpha1,name=vvolumereplicationgroup.kb.io,admissionReviewVersions=v1
+
+func vrgValidatorWebhookRegister(mgr manager.Manager) {
+	mgr.GetWebhookServer().Register(
+		"/validate-ramendr-openshift-io-v1alpha1-volumereplicationgroup",
+		&admission.Webhook{Handler: &vrgValidator{
+			client:  mgr.GetClient(),
+			log:     mgr.GetLogger().WithName("webhook").WithName("vrg").WithName("validate"),
+			decoder: func() *admission.Decoder { d, _ := admission.NewDecoder(mgr.GetScheme()); return d }(), //nolint:errcheck,nlreturn,lll
+			// TODO fix with controller-runtime v0.15+
+			// decoder: admission.NewDecoder(mgr.GetScheme()),
+		}},
+	)
+}
+
+type vrgValidator struct {
+	client  client.Client
+	log     logr.Logger
+	decoder *admission.Decoder
+}
+
+var _ admission.Handler = &vrgValidator{}
+
+func (v *vrgValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := v.log
+
+	vrg := &ramen.VolumeReplicationGroup{}
+	if err := v.decoder.Decode(req, vrg); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	vrgNamespacedName := types.NamespacedName{
+		Namespace: vrg.Namespace,
+		Name:      vrg.Name,
+	}
+	log = log.WithValues("vrg", vrgNamespacedName.String())
+
+	if vrg.Spec.KubeObjectProtection == nil ||
+		vrg.Spec.KubeObjectProtection.RecipeRef == nil {
+		return admission.Allowed("recipe not specified")
+	}
+
+	recipeNamespacedName := types.NamespacedName{
+		Namespace: vrg.Spec.KubeObjectProtection.RecipeRef.Namespace,
+		Name:      vrg.Spec.KubeObjectProtection.RecipeRef.Name,
+	}
+
+	recipe := &recipe.Recipe{}
+	if err := v.client.Get(ctx, recipeNamespacedName, recipe); err != nil {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("recipe %#v retrieval error: %w", recipeNamespacedName, err))
+	}
+
+	if err := RecipeParametersExpand(recipe, vrg.Spec.KubeObjectProtection.RecipeParameters, log); err != nil {
```

Comment:
RecipeParameters can be expanded to add more namespaces later by the user, which would be a problem?